File: src/main/java/com/lannstark/lec12/Lec12Main.java
package com.lannstark.lec12;

import yoon.lec12.Person;

public class Lec12Main {

  public static void main(String[] args) {
    Person.Factory.newBaby("윤영진");
//    Person.Companion.newBaby("윤영진");
//    Person.newBaby("윤영진");
    moveSomething(new Movable() {
      @Override
      public void move() {

      }

      @Override
      public void fly() {

      }
    });

  }

  private static void moveSomething(Movable movable) {
    movable.move();
    movable.fly();
  }

}

File: src/main/kotlin/yoon/lec12/Lec12Main.kt
package yoon.lec12

import com.lannstark.lec12.Movable

fun main() {
//    println(Person.newBaby("윤영진"))
//    Singleton.a
    moveSomething(object : Movable {
        override fun move() {
        }

        override fun fly() {
        }

    })
}

private fun moveSomething(movable: Movable) {
    movable.move()
    movable.fly()
}

class Person constructor(
    private var name: String,
    private var age: Int,
) {
    companion object Factory : Log {
        private const val MIN_AGE = 1

        @JvmStatic
        fun newBaby(name: String): Person {
            return Person(name, MIN_AGE)
        }

        override fun log() {
            println("log ...")
        }
    }
}

